src/rust/inetstack/protocols/layer4/tcp/established/receiver.rs

```diff
@@ -274,60 +186,166 @@ impl Receiver {
             }
         };
 
-        // Check whether we've received the last packet in this TCP stream.
+        // Have we received all data before the FIN?
         if cb
             .receiver
             .fin_seq_no
             .get()
             .is_some_and(|seq_no| seq_no == cb.receiver.receive_next_seq_no)
         {
-            // Once we know there is no more data coming, begin closing down the connection.
-            Self::process_fin(cb);
+            let state = match cb.state {
+                State::Established => State::CloseWait,
+                State::FinWait1 => State::Closing,
+                State::FinWait2 => State::TimeWait,
+                state => unreachable!("Cannot be in any other state at this point: {:?}", state),
+            };
+            cb.state = state;
+            cb.receiver.pop_queue.push(DemiBuffer::new(0));
+            debug_assert_eq!(cb.receiver.receive_next_seq_no, cb.receiver.fin_seq_no.get().unwrap());
+            // Reset it to wake up any close coroutines waiting for FIN to arrive.
+            cb.receiver.fin_seq_no.set(Some(cb.receiver.receive_next_seq_no));
+            // Move RECV_NXT over the FIN.
+            cb.receiver.receive_next_seq_no = cb.receiver.receive_next_seq_no + 1.into();
         }
 
-        // Send an ack on every FIN. We do this separately here because if the FIN is in order, we ack it after the
-        // previous line, otherwise we do not ack the FIN.
+        // Have we processed all of the data and the FIN?
         if header.fin {
-            Sender::send_ack(cb, layer3_endpoint)
+            if cb.state == State::Closing {
+                warn!("Simultaneous close detected. Resending FIN.");
+                Sender::send_fin(cb, layer3_endpoint, now)?;
+            } else {
+                Sender::send_ack(cb, layer3_endpoint);
+            }
         }
 
-        // We should ACK this segment, preferably via piggybacking on a response.
-        if cb.receiver.ack_deadline_time_secs.get().is_none() {
-            // Start the delayed ACK timer to ensure an ACK gets sent soon even if no piggyback opportunity occurs.
-            let timeout: Duration = cb.receiver.ack_delay_timeout_secs;
-            // Getting the current time is extremely cheap as it is just a variable lookup.
-            cb.receiver.ack_deadline_time_secs.set(Some(now + timeout));
-        } else if has_data {
-            // We already owe our peer an ACK (the timer was already running), so cancel the timer and ACK now.
-            cb.receiver.ack_deadline_time_secs.set(None);
-            trace!("process_packet(): sending ack before deadline because another packet arrived");
-            Sender::send_ack(cb, layer3_endpoint);
+        Ok(())
+    }
+
+    pub fn get_receive_window_size(&self) -> u32 {
+        let bytes_unread: u32 = (self.receive_next_seq_no - self.reader_next_seq_no).into();
+        // The window should be less than 1GB or 64KB without scaling.
```

Comment:
It would be good to not assume these absolute sizes in the comments. These are susceptible to going stale in case the values of constants change in future.

Reply:
Oh, these are TCP constants, so won't change unless something big changes in the RFC.